keps/sig-auth/541-external-credential-providers/README.md

```diff
@@ -127,24 +207,129 @@ be readable and executable by the client process.
+  Spec ExecCredentialSpec `json:"spec,omitempty"`
+
+  // Status is filled in by the plugin and holds the credentials that the
+  // transport should use to contact the API.
+  // +optional
+  Status *ExecCredentialStatus `json:"status,omitempty"`
+}
+
+// ExecCredentialSpec holds request and runtime specific information provided by
+// the transport.
+type ExecCredentialSpec struct {
+  // Cluster contains information to allow an exec plugin to communicate with the
+  // kubernetes cluster being authenticated to.
+  Cluster Cluster `json:"cluster"`
+}
+
+// Cluster contains information to allow an exec plugin to communicate with the
+// kubernetes cluster being authenticated to.
+type Cluster struct {
+  // Server is the address of the kubernetes cluster (https://hostname:port).
+  Server string `json:"server"`
+  // ServerName is passed to the server for SNI and is used in the client to
+  // check server certificates against. If ServerName is empty, the hostname
+  // used to contact the server is used.
+  // +optional
+  ServerName string `json:"serverName,omitempty"`
+  // CAData contains PEM-encoded certificate authority certificates.
+  // If empty, system roots should be used.
+  // +listType=atomic
+  // +optional
+  CAData []byte `json:"caData,omitempty"`
+  // Config holds additional config data that is specific to the exec plugin
+  // with regards to the cluster being authenticated to.
+  // +optional
+  Config runtime.RawExtension `json:"config,omitempty"`
+}
+```
+
+The Go struct for the `clusters.[...].cluster.extensions[...].extension` field:
+
+```golang
+// Cluster contains information about how to communicate with a kubernetes cluster
+type Cluster struct {
+  // Server is the address of the kubernetes cluster (https://hostname:port).
+  Server string `json:"server"`
+
+  ... omitted for brevity ...
+
+  // Extensions holds additional information. This is useful for extenders so
+  // that reads and writes don't clobber unknown fields
+  // +optional
+  Extensions []NamedExtension `json:"extensions,omitempty"`
+}
 
+// NamedExtension relates nicknames to extension information
+type NamedExtension struct {
+  // Name is the nickname for this Extension
+  Name string `json:"name"`
+  // Extension holds the extension information
+  Extension runtime.RawExtension `json:"extension"`
+}
 ```
+
+The `spec.cluster` field is the current cluster that the client is communicating
+with (i.e it is the cluster the client knows it must communicate with after it
+has completed parsing its `kubeconfig`, flags and environment variables).
+This allows the executable to perform different actions based on the current
+cluster (i.e. get a token for a particular cluster).  The `Cluster` struct is
+flexible in that it not only provides all details required to communicate with
+the cluster (hostname and TLS config), but that is also allows arbitrary
+per-cluster configuration to be passed to the executable via the `config` field.
```

Comment:
I can see use cases for cluster hostname and custom config bits, but I was surprised by the ServerName and CAData bits... are we expecting exec plugins to call out to the server we are asking them to provide credentials for?

If so, don't we need to include ProxyURL and InsecureSkipTLSVerify?

Reply:
Opened a new PR into the KEP to try to address this design update: https://github.com/kubernetes/enhancements/pull/1927.